pkg/image/save/oci_save.go

```diff
@@ -0,0 +1,193 @@
+		} else if is.isImageExist(skopeo.ContainersStorage, named, "") {
+			is.typeToImages[skopeo.ContainersStorage] = append(is.typeToImages[skopeo.ContainersStorage], named)
+		} else {
+			is.typeToImages[skopeo.RemoteRegistry] = append(is.typeToImages[skopeo.RemoteRegistry], named)
+		}
+		progress.Message(is.progressOut, "", fmt.Sprintf("Pulling image: %s", named.FullName()))
+	}
+
+	//perform image save ability
+	eg, _ := errgroup.WithContext(context.Background())
+	numCh := make(chan struct{}, maxPullGoroutineNum)
+	for imageType, nameds := range is.typeToImages {
+		tmpnameds := nameds
+		tmpType := imageType
+		numCh <- struct{}{}
+		eg.Go(func() error {
+			defer func() {
+				<-numCh
+			}()
+			err := is.save(tmpType, tmpnameds, "", dir)
+			if err != nil {
+				return fmt.Errorf("ImageType:%s failed to save domain %s image: %v", imageType, tmpnameds[0].domain, err)
+			}
+			return nil
+		})
+	}
+	if err := eg.Wait(); err != nil {
+		return err
+	}
+	if len(images) != 0 {
+		progress.Message(is.progressOut, "", "Status: images save success")
+	}
+	return nil
+}
+
+func (is *OCIImageSaver) SaveImagesWithAuth(imageList ImageListWithAuth, dir string, platform v1.Platform) error {
+	//init a pipe for display pull message
+	reader, writer := io.Pipe()
+	defer func() {
+		_ = reader.Close()
+		_ = writer.Close()
+	}()
+	is.progressOut = streamformatter.NewJSONProgressOutput(writer, false)
+	is.ctx = context.Background()
+	go func() {
+		err := dockerjsonmessage.DisplayJSONMessagesToStream(reader, dockerstreams.NewOut(common.StdOut), nil)
+		if err != nil && err != io.ErrClosedPipe {
+			logrus.Warnf("error occurs in display progressing, err: %s", err)
+		}
+	}()
+
+	//perform image save ability
+	eg, _ := errgroup.WithContext(context.Background())
+	numCh := make(chan struct{}, maxPullGoroutineNum)
+
+	//handle imageList
+	for _, section := range imageList {
+		for _, nameds := range section.Images {
+			tmpnameds := nameds
+			creds := section.Username + ":" + section.Password
+			progress.Message(is.progressOut, "", fmt.Sprintf("Pulling image: %s", tmpnameds[0].FullName()))
+			numCh <- struct{}{}
+			eg.Go(func() error {
+				defer func() {
+					<-numCh
+				}()
+				if err := is.save(skopeo.RemoteRegistry, tmpnameds, creds, dir); err != nil {
+					return err
+				}
+				return nil
+			})
+		}
+		if err := eg.Wait(); err != nil {
+			return err
+		}
+	}
+
+	if len(imageList) != 0 {
+		progress.Message(is.progressOut, "", "Status: images save success")
+	}
+	return nil
+}
+
+func (is *OCIImageSaver) save(imageType string, namds []Named, creds string, dir string) error {
+	for _, namd := range namds {
+		srcImageName := getImageNameWithStorageType(imageType, namd, "")
+		destImageName := getImageNameWithStorageType(skopeo.OciPath, namd, dir)
+		if err := getDestDir(destImageName); err != nil {
+			return err
+		}
+		err := skopeo.Copy(srcImageName, destImageName, creds)
```

Comment:
This approach will make it impossible to reuse layers that may be duplicated in each image while building new oci image.